### PR TITLE
[FIX] mrp_account: adds business domain to bom check analytic

### PR DESCRIPTION
**Pull request by yosa-odoo**
Steps to reproduce:
-Go to analytic plans -> Projects -> Add a new line with a "Purchase Order" Domain and set Applicability to "Mandatory" -> Save
-Go to Bills oF Materials -> create a new one w/ any product -> Save

Issue:
-There will be an error "One or more lines require a 100% analytic distribution."

Cause:
When saving the BoM, it leads to a constraint check
https://github.com/odoo/odoo/blob/0217b88e481dc1d0d244b620ba527dec177edde1/addons/mrp_account/models/mrp_bom.py#L44-L50

In this method, we first try to find the relevant plans:
https://github.com/odoo/odoo/blob/d6bdb05771fd29a79f2e8087b92a438fec28afaa/addons/analytic/models/analytic_mixin.py#L103-L107

To do so, we take all existing plans, evaluate them with a score, and take the one with the higher score. Here is the issue: when evaluating *Projects*, since we didn't provide any business domain and since the plan has a company, we assign a 0.5 score to the plan instead of skipping it:
https://github.com/odoo/odoo/blob/f17b37b14cad8f353443586cc9c4ecaf86e476da/addons/analytic/models/analytic_plan.py#L310-L319
Which does not make sense as the business domain of *Project* is *Purchase Order* and here the user is dealing with a BoM...

As a result, back to `_validate_distribution`, we check the distribution for this plan and realize there isn't any, hence the error:
https://github.com/odoo/odoo/blob/d6bdb05771fd29a79f2e8087b92a438fec28afaa/addons/analytic/models/analytic_mixin.py#L115-L117

Solution:

BOM: We don't restrict anymore the creation of a BOM. Setting an analytic distribution is only a tool to facilitate the definition of the analytic distribution on the MO but the creation should not be constrained
Workcenter: we correct the flowing by specifying the business domain restricting the creation of a workcenter if any relevant rule is present

### CODE REVIEW

File: addons/mrp_account/tests/test_analytic_account.py
Comment (adwid) [inline addons/mrp_account/tests/test_analytic_account.py:450]: The error message does not provide any relevant info here, the code is clear enough
```suggestion
```
Same with almost all next assert actually (I would say that only the msg for the `assertRaises` might be useful)

**Review by adwid — COMMENTED** [review]
I agree with your fix

Thx for the tests, just a small nitpicking for the asserts

About the commit redaction, I think it is clear for people who are used to work with analytics only. So you could make a bit more clear for everybody ? Like:

When saving the BoM, it leads to a constraint check
https://github.com/odoo/odoo/blob/0217b88e481dc1d0d244b620ba527dec177edde1/addons/mrp_account/models/mrp_bom.py#L44-L50

In this method, we first try to find the relevant plans:
https://github.com/odoo/odoo/blob/d6bdb05771fd29a79f2e8087b92a438fec28afaa/addons/analytic/models/analytic_mixin.py#L103-L107

To do so, we take all existing plans, evaluate them with a score, and take the one with the higher score. Here is the issue: when evaluating *Projects*, since we didn't provide any business domain and since the plan has a company, we assign a 0.5 score to the plan instead of skipping it:
https://github.com/odoo/odoo/blob/f17b37b14cad8f353443586cc9c4ecaf86e476da/addons/analytic/models/analytic_plan.py#L310-L319
Which does not make sense as the business domain of *Project* is *Purchase Order* and here the user is dealing with a BoM...

As a result, back to `_validate_distribution`, we check the distribution for this plan and realize there isn't any, hence the error:
https://github.com/odoo/odoo/blob/d6bdb05771fd29a79f2e8087b92a438fec28afaa/addons/analytic/models/analytic_mixin.py#L115-L117

**Review by adwid — APPROVED** [review]
Side note: constraining the BoM does not make sense, we do it later on the MO, which is more logic:
https://github.com/odoo/odoo/blob/0ff125b9573ae0977bed430360cbd3228380008f/addons/mrp_account/models/mrp_production.py#L40-L46

**Comment by yosa-odoo** [comment]
discussed with @mathieu-mgm

**Comment by adwid** [comment]
Well, the diff is quite the same
You just removed one constraint, following a discussion with Matthieu
So I still agree

**Comment by yosa-odoo** [comment]
Top, I'll wait for the review from logistics!

**Comment by amoyaux** [comment]
robodoo r+

**Pull request closed without merging**
